lib/rights_api/query_optimizer.rb

```diff
@@ -0,0 +1,73 @@
+# frozen_string_literal: true
+
+require "digest/md5"
+require "json/canonicalization"
+
+require_relative "services"
+
+module RightsAPI
+  class QueryOptimizer
+    attr_reader :parser, :offset, :where
+
+    # A class that translates parsed query and optimizes it to
+    # avoid large OFFSETS into large tables when previous (cached) queries
+    # may enable the use of smaller or zero OFFSET.
+    # Doing this requires a model class that returns `true` for `#optimize?`
+    # and implements `.optimizer_query`, the results of which are used to derive
+    # additional `where` clause(s) that reduce or eliminate the OFFSET.
+    #
+    # This is only used with rights_current and rights_log.
+    #
+    # Provides a modified `#offset` that supersedes the query parser's version,
+    # and `#where` clause(s) that should be applied after the parser's `#where`
+    # that should replace or augment the desired OFFSET.
+    def initialize(parser:)
+      @parser = parser
+      @where = []
+      @offset = parser.offset
+      optimize
+    end
+
+    def add_to_cache(dataset:)
+      return if ENV["RIGHTS_API_DISABLE_OFFSET_OPTIMIZER"]
+      return if dataset.count.zero?
+      return unless dataset.last.class.optimize?
+
+      data = {
+        offset: parser.offset,
+        limit: parser.limit,
+        last: dataset.last.to_h
+      }
+      Services[:cache][model_params_md5] = data
+    end
+
+    private
+
+    def optimize
+      return if ENV["RIGHTS_API_DISABLE_OFFSET_OPTIMIZER"]
+      return if offset.zero?
+
+      cached = Services[:cache][model_params_md5]
+      return if cached.nil?
```

Comment:
If I'm reading this correctly, if somebody happens to hit a different instance of the rights API (e.g. running in another pod or another datacenter) then the optimizer won't be used because the optimizer won't have the mapping from offset to timestamp cached, correct?